lib/ntbb-session.lib.php

```diff
@@ -202,6 +202,18 @@ private function passwordVerifyInner($userid, $pass, $user) {
 			}
 		}
 
+		if (substr(@$user['email'], -1) === '@') {
+			require_once dirname(__FILE__).'/../vendor/autoload.php';
+			$client = new Google_Client(['client_id' => $psconfig['gapi_clientid']]);
+			$payload = $client->verifyIdToken($pass);
+			if (!$payload) return false;
+			if (strpos($payload['aud'], $psconfig['gapi_clientid']) === false) return false;
+			if ($payload['email'] === substr($user['email'], 0, -1)) {
+				return true;
+			}
+			return false;
+		}
+
 		$rehash = false;
 		if ($user['passwordhash']) {
```

Comment:
If Google Login is enforced for an account, we no longer need to store the password hash, and it should probably be deleted, least an attacker who vulnerates the login server retrieves it by targetting these specially marked accounts and building a rainbow table from the hash.

Reply:
Speaking of login server being taken over...

To make sure we are all in the same page, it seems that 2FA wouldn't improve security after that, since with this implementation as long as they get the login private key, they can fake a Google login without ever speaking to Google servers.

Given that, it would be helpful to be clear of the threat model we are working under, to make sure that our 2FA implementation really does address menaces.

Reply:
We don't check the password hash if Google login is enforced, though...

Reply:
The lessons from the past (rootinabox and last year's incident) would say that the real-world security threat lies in he login server itself. Therefore,  would it be possible to validate Google logins directly in the sim server?

IMO, Ideally a 2FA implementation would require *independent* login assertions from psweb and gapi. However, it seems you want to rely only on Google's own 2FA implementation.

Reply:
I suppose we _could_? The purpose of this commit is to protect the login server, though, which is way more of an issue than the sim server.

Reply:
I don't see what's wrong with relying on Google's own 2FA implementation? You know how 2FA works, right? Adding something else you know and something else you have doesn't actually make it any more secure.

Reply:
Like, validating logins in the sim server is like going through airport security after the plane has landed: it's not _completely_ useless, but it's not the main thing we're trying to protect against.

Reply:
"The lessons from the past is that people keep robbing all the cash from our bank's safes."

"We should get a huge lock..."

"and?"

"and put the huge lock on our break room, so bank robbers can't get into our break room after taking all the money out of our safes."

Reply:
I don't see what you are arguing for. Analogies are bad, for certain.
How does this protect the login server? Again, what's our threat model?

Reply:
> We don't check the password hash if Google login is enforced, though...

Just read this. Indeed, but we store it, and if the hashes get exposed they can be eventually(TM) broken to access other sites with the same credentials.

The risk is minimal, but it's still there, and it affects the users.

Reply:
Our threat model is "someone steals a password and uses it to log into a PS admin account".

Reply:
Seems niche, but as long as you assert that Google credentials are somehow (without the "know and own" condition?!) inherently safer than PS credentials, I guess this works.

Reply:
The Google credentials come with the "know and own" conditions...

Reply:
Like, the threat models you're talking about aren't really important? I've never seen Google worry about them, and they take security very seriously.

Reply:
Also: There's nothing in this code that means we have to store password hashes in Google login mode.

Reply:
> The Google credentials come with the "know and own" conditions...

But they are opt-in for the user. Unless I am missing something?

> I've never seen Google worry about them.
Of course they are really important. Otherwise our security efforts are firing without aim.
From a quick Google search, in the abstract of Chromium OS sec overview.
https://sites.google.com/a/chromium.org/dev/chromium-os/chromiumos-design-docs/security-overview

Reply:
> But they are opt-in for the user. Unless I am missing something?

Yes, they're opt-in for the user. Why is that a problem? _Most_ two-factor auth is opt-in for the user.

> Of course they are really important. Otherwise our security efforts are firing without aim.

The page you link to makes no mention of the threat models you've been talking about.

Reply:
Once again:

> The lessons from the past (rootinabox and last year's incident) would say that the real-world security threat lies in the login server itself. Therefore, would it be possible to validate Google logins directly in the sim server?

Our biggest threat model is "attacker takes down the login server".

Protecting the sim server against the login server is **useless** because **our goal is to protect the login server**. There is nothing valuable in the sim server, everything of value is in the login server. The sim server can't do anything without a login server anyway!

Reply:
The problem is that, as you have stated, we care specifically about admin credentials, and this:
- Is unsafe by default, by which I mean "as safe as it already is", unless -again- Google authentication process is otherwise somehow safer than ours', an idea with which I can live, but should be stated clearly.
- Relies on third party authentication methods.
- Empowers a third party which may act as a opportunistic adversary in the adminship context (Would be fixed by still requiring the PS password)

Reply:
> There is nothing valuable in the sim server, everything of value is in the login server.

Years worth of modlogs and chatlogs. Quite valuable probably.
We can probably dismiss the trivia questions DB (is there even a command to dump it?), as well as the dumpable data stored sim-server-side for all side servers.

Reply:
We can't _force_ admins to be safe, but we've never been able to. Our policy will be to tell admins to enable 2FA on Google, maybe require screenshots of 2FA being enabled, and that will always be the best we can do. Policies that make it harder for admins to circumvent things like this are of very marginal value.

Relying on third-party authentication methods is a good thing when the third party is Google. They say "don't roll your own crypto" and I tend to generally agree with that.

I also don't consider "Google wants to take over PS" as a particularly worrying threat model. I suppose it's something I'm willing to live with.

Reply:
Are they valuable though? Chat logs and modlogs probably aren't overly important imo. Regular users can store chat logs and I don't see why a hacker would ever want the modlogs. I suppose the dumpable data from side servers could be an issue though

Reply:
If your recovery email is a Gmail account, you already trust Google with basically everything. And I'd argue this is a good thing – I trust Google to secure an email server much more than I trust myself or any other company to.

Reply:
> Are they valuable though? Chat logs and modlogs probably aren't overly important imo. Regular users can store chat logs and I don't see why a hacker would ever want the modlogs.

I am sure you could get a few zeros from mining/selling PS logs.

> If your recovery email is a Gmail account, you already trust Google with basically everything. And I'd argue this is a good thing – I trust Google to secure an email server much more than I trust myself or any other company to.

I barely rely on Google/Gmail myself, but I can see where you are coming from.

> They say "don't roll your own crypto" and I tend to generally agree with that.

This isn't "our own crypto". We use bcrypt, don't we?

> We can't force admins to be safe, but we've never been able to. Our policy will be to tell admins to enable 2FA on Google, maybe require screenshots of 2FA being enabled, and that will always be the best we can do. Policies that make it harder for admins to circumvent things like this are of very marginal value.
Relying on third-party authentication methods is a good thing when the third party is Google. They say "don't roll your own crypto" and I tend to generally agree with that.
I also don't consider "Google wants to take over PS" as a particularly worrying threat model. I suppose it's something I'm willing to live with.

With that out of the way, lgtm.

Reply:
Oh I suppose that's possible, I'm totally forgetting about the recent PSRevealer incident

Reply:
The sim server's logs are valuable, but nowhere near as valuable as the data in the login server.

I'm not opposed to adding more checks in the sim server, but for now I consider this pullreq a strict improvement in security, barring, like, actual vulns.

Reply:
Please also remember that the login server data available to admins is the same as to leaders (at least that was the case last time I checked /users panel as admin).

Just for awareness sake.

(Of course, password reset access is very important; however, that doesn't imply data access per se, but in conjunction with the logs/sim server)